src/Compilers/CSharp/Portable/BoundTree/TupleBinaryOperatorInfo.cs

```diff
@@ -0,0 +1,113 @@
+        {
+            internal readonly BinaryOperatorKind Kind;
+            internal readonly Conversion LeftConversion;
+            internal readonly Conversion RightConversion;
+            internal readonly MethodSymbol MethodSymbolOpt; // User-defined comparison operator, if applicable
+
+            internal readonly Conversion ConversionForBool; // If a conversion to bool exists, then no operator needed. If an operator is needed, this holds the conversion for input to that operator.
+            internal readonly UnaryOperatorSignature BoolOperator; // Information for op_true or op_false
+
+            internal Single(TypeSymbol leftConvertedType, TypeSymbol rightConvertedType, BinaryOperatorKind kind,
+                Conversion leftConversion, Conversion rightConversion, MethodSymbol methodSymbolOpt,
+                Conversion conversionForBool, UnaryOperatorSignature boolOperator) : base(leftConvertedType, rightConvertedType)
+            {
+                Kind = kind;
+                LeftConversion = leftConversion;
+                RightConversion = rightConversion;
+                MethodSymbolOpt = methodSymbolOpt;
+                ConversionForBool = conversionForBool;
+                BoolOperator = boolOperator;
+
+                Debug.Assert(Kind.IsUserDefined() == ((object)MethodSymbolOpt != null));
+            }
+
+            internal override bool IsSingle()
+                => true;
+
+            public override string ToString()
+                => $"binaryOperatorKind: {Kind}";
+
+#if DEBUG
+            internal override TreeDumperNode DumpCore()
+            {
+                var sub = new List<TreeDumperNode>();
+                if ((object)MethodSymbolOpt != null)
+                {
+                    sub.Add(new TreeDumperNode("methodSymbolOpt", MethodSymbolOpt.ToDisplayString(), null));
+                }
+                sub.Add(new TreeDumperNode("leftConversion", LeftConvertedType.ToDisplayString(), null));
+                sub.Add(new TreeDumperNode("rightConversion", RightConvertedType.ToDisplayString(), null));
+
+                return new TreeDumperNode("nested", Kind, sub);
+            }
+#endif
+        }
+
+        /// <summary>
+        /// Holds the information for a tuple comparison, either at the top-level (like `(a, b) == ...`) or nested (like `(..., (a, b)) == (..., ...)`).
+        /// </summary>
+        internal class Multiple : TupleBinaryOperatorInfo
+        {
+            internal readonly ImmutableArray<TupleBinaryOperatorInfo> Operators;
+
+            internal Multiple(ImmutableArray<TupleBinaryOperatorInfo> operators, TypeSymbol leftConvertedType, TypeSymbol rightConvertedType)
+                : base(leftConvertedType, rightConvertedType)
+            {
+                Debug.Assert(leftConvertedType.StrippedType().IsTupleType);
+                Debug.Assert(rightConvertedType.StrippedType().IsTupleType);
+                Debug.Assert(!operators.IsDefault);
```

Comment:
>!operators.IsDefault [](start = 29, length = 20)

Is an empty set a valid input to the constructor? It feels like there must be at least two items in the array.

Reply:
I see, empty array is used for cardinality mismatch cases. Consider asserting that otherwise we have at least two items.

---
In reply to: [172088222](https://github.com/dotnet/roslyn/pull/23356#discussion_r172088222) [](ancestors = 172088222)